pkg/registry/core/service/storage/storage.go

```diff
@@ -241,6 +245,63 @@ func (r *GenericREST) defaultOnReadService(service *api.Service) error {
 			}
 		}
 	}
+}
+
+func (r *GenericREST) beginCreate(ctx context.Context, obj runtime.Object, options *metav1.CreateOptions) (func(context.Context, bool), error) {
+	svc := obj.(*api.Service)
+
+	klog.Errorf("TIM: beginCreate: svc %v/%v - prepare to allocate", svc.Namespace, svc.Name)
+
+	// Make sure ClusterIP and ClusterIPs are in sync.  This has to happen
+	// early, before anyone looks at them.
+	// NOTE: the args are (old, new)
+	svcreg.NormalizeClusterIPs(nil, svc)
+
+	// Allocate IPs and ports. If we had a transactional store, this would just
+	// be part of the larger transaction.  We don't have that, so we have to do
+	// it manually. This has to happen here and not in any earlier hooks (e.g.
+	// defaulting) because it needs to be aware of flags and be able to access
+	// API storage.
+	txn, err := r.alloc.allocateCreate(svc, dryrun.IsDryRun(options.DryRun))
+	if err != nil {
+		return nil, err
+	}
+
+	// Our cleanup callback
+	finish := func(_ context.Context, success bool) {
```

Comment:
might be better to replace success with error variable. Allowing finish functions to log error and call commit/revert. It will show user `this thing was reverted because the op generated this xyz error`

Reply:
It originally did, but that made the generic code more complex wrt variable capture.  If API machinery folks want it back, I am happy to do so.